python/feathub/metric_stores/metric.py

```diff
@@ -192,3 +210,150 @@ def from_json(cls, json_dict: Dict) -> "Ratio":
+                            since the Feathub job is created.
+        """
+        super().__init__("count_map", window_size)
+
+    def get_tags(self) -> OrderedDict:
+        tags = super(CountMap, self).get_tags()
+        tags["count_map_key"] = "null"
+        return tags
+
+    def _get_transform_functions(self) -> Sequence[Callable[[str], Transformation]]:
+        return [
+            lambda feature_name: SlidingWindowTransform(
+                expr=f"CAST({feature_name} AS STRING)",
+                agg_func="VALUE_COUNTS",
+                window_size=self.window_size,
+                step_size=self.window_size,
+            ),
+        ]
+
+    @append_metadata_to_json
+    def to_json(self) -> Dict:
+        return {
+            "window_size_sec": self.window_size / timedelta(seconds=1),
+        }
+
+    @classmethod
+    def from_json(cls, json_dict: Dict) -> "CountMap":
+        return CountMap(
+            window_size=timedelta(seconds=json_dict["window_size_sec"]),
+        )
+
+
+class RatioMap(Metric):
+    """
+    Metrics that show the proportion of occurrence of each feature value.
+    """
+
+    def __init__(self, window_size: timedelta = timedelta(seconds=0)):
+        """
+        :param window_size: The time range to compute the metric. It should be zero or
+                            a positive time span. If it is zero, the metric will be
+                            computed from all feature values that have been processed
+                            since the Feathub job is created.
+        """
+        super(RatioMap, self).__init__("ratio_map", window_size)
+
+    def get_tags(self) -> OrderedDict:
+        tags = super(RatioMap, self).get_tags()
+        tags["ratio_map_key"] = "null"
+        return tags
+
+    def get_transform_functions_and_output_types(
+        self,
+    ) -> Sequence[
+        Tuple[Callable[[str], Transformation], Callable[[DType], Optional[DType]]]
+    ]:
+        def count_to_ratio(
+            row: pd.Series, feature_name: str
+        ) -> Optional[Dict[str, float]]:
+            ratio_map: Dict[str, float] = row[feature_name]
+            if isinstance(ratio_map, np.float64) and np.isnan(ratio_map):
+                # TODO: figure out why map features whose value is None would be
+                #  converted to np.nan before passing to PyFlink UDF
+                return None
+            count = sum(ratio_map.values())
+            for key in ratio_map.keys():
+                ratio_map[key] /= count
+            return ratio_map
+
+        return [
+            (
+                lambda feature_name: SlidingWindowTransform(
+                    expr=f"CAST({feature_name} AS STRING)",
+                    agg_func="VALUE_COUNTS",
+                    window_size=self.window_size,
+                    step_size=self.window_size,
+                ),
+                lambda dtype: None,
+            ),
+            (
+                lambda feature_name: PythonUdfTransform(
```

Comment:
It might be more intuitive to let `Transformation` specify its output data type, instead of making back-and-forth changes to the metric API.

And it seems simpler not to support RatioMap if it is easy for users to derive it by themselves in Prometheus.